### PR TITLE
fix: Pass `--platform=manylinux2014` to `pip install`

### PR DESCRIPTION
**Pull request by pdecat**
## Description

As mentioned in [AWS documentation](https://repost.aws/knowledge-center/lambda-python-package-compatible), the `--platform=manylinux2014` parameter must be passed to `pip install` for versions of packages with compiled binaries  compatible with Lambda runtimes to be installed.

Not doing so causes wrong versions of python wheel packages to be installed when applying a terraform configuration with Lambda resources on platforms with a newer glibc than Lambda supports.

This results in runtime errors like this one with recent cryptography packages:
```
/lib64/libc.so.6: version `GLIBC_2.28' not found (required by /opt/python/cryptography/hazmat/bindings/_rust.abi3.so
```

## Motivation and Context

Resolves https://github.com/terraform-aws-modules/terraform-aws-lambda/issues/346

### For reference:

* `python3.7` runtime uses Amazon Linux 1 with glibc 2.17
* `python3.8` and `python3.9` runtime use Amazon Linux 2 with glibc 2.26
* `manylinux2014` supports glibc 2.17

https://repost.aws/knowledge-center/lambda-python-package-compatible
https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html
https://aws.amazon.com/fr/amazon-linux-2/faqs/
https://peps.python.org/pep-0599/#the-manylinux2014-policy
https://github.com/pyca/cryptography/issues/6391

## Breaking Changes

No.

## How Has This Been Tested?
- [ ] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [ ] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by pdecat** [comment]
Marked as draft as testing is still in progress.

**Comment by pdecat** [comment]
CI seems broken because of network issues:

```
│ Error: Failed to install provider
│ 
│ Error while installing hashicorp/external v2.3.1: read tcp
│ 10.1.1.20:40466->108.138.85.65:443: read: connection reset by peer
```

**Comment by antonbabenko** [comment]
@pdecat Rerun helped.

File: package.py [hunk 1]
```diff
@@ -980,10 +980,16 @@ def install_pip_requirements(query, requirements_file, tmp_dir):
...
-                'install', '--no-compile',
-                '--prefix=', '--target=.',
-                '--requirement={}'.format(requirements_filename),
+                python_exec,
+                "-m",
+                "pip",
+                "install",
+                "--no-compile",
+                "--only-binary=:all:",
+                "--platform=manylinux2014",
```
Comment (pdecat) [inline package.py:989]: Passing `--platform=manylinux2014` requires either `--only-binary=:all:` or `--no-deps`.

**Comment by pdecat** [comment]
I guess I'll close this PR, and use the aforementioned work-around until Amazon releases support for Python 3.10, or better yet 3.11, with runtimes based on [Amazon Linux 2023 that comes with glibc 2.34](https://docs.aws.amazon.com/linux/al2023/ug/compare-with-al2.html#glibc-gcc-and-binutils).

**Comment by pdecat** [comment]
After thinking a bit more about it, I believe this can be achieved by:
* passing the value of the function `architectures` or layer `compatible_architectures` in the `query` of the `data.external.archive_prepare` datasource that invokes `package.py prepare`
* then only pass `--only-binary=:all: --platform=manylinux2014_{architecture}` to pip install if `architectures` is defined and contains a single value.

This should avoid breaking any existing usages of the module.
And to opt-in the new behavior, one would have to define a single architecture for its function or layer.

**Comment by antonbabenko** [comment]
@pdecat Thank you very much for the work you do on this module!

I think, I like the workaround described in https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/444#issuecomment-1494592708 very much because all user will have to do is just set a couple of environment variables and we avoid putting extra logic into already big `package.py`.

If it is documented properly, it can be a decent solution. What do you think?

**Comment by pdecat** [comment]
> @pdecat Thank you very much for the work you do on this module!

Thanks @antonbabenko, I'm glad to contribute to your awesome modules :)

> I think, I like the workaround described in [#444 (comment)](https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/444#issuecomment-1494592708) very much because all user will have to do is just set a couple of environment variables and we avoid putting extra logic into already big `package.py`.
> 
> If it is documented properly, it can be a decent solution. What do you think?

I do agree `package.py` is already relatively complex, but I believe my [last proposed change](https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/444#issuecomment-1495627793) should not be too intrusive or risky (in the sense of breaking existing configurations).

IMO, the work-around using environment variables can be acceptable, but it would be better if their presence could be enforced one way or another.
Indeed, one thing that I find less desirable about using environment variables is that they are not taken into account when computing the package hash sum, which is key to detect something as changed.
Maybe using [pip configuration files](https://pip.pypa.io/en/stable/topics/configuration/#configuration-files) instead of environments variables could help get a better work-around as they could probably be included in the hash computation using `source_path`.

Also, what's your take on [restoring `hash_extra_paths` behavior](https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/444#issuecomment-1494359864), maybe also allowing the user to customize it by exposing it as variable with a default value? Should I open a new issue to discuss that matter further?

**Pull request closed without merging**